### PR TITLE
Make `SourceInfo::get_next_message` a lot simpler

### PR DESCRIPTION
**Pull request by ruchirK**
In particular - make the implementer responsible for deduplicating records, and give them a way to indicate "there was a temporary glitch please check again later"

I think the Kinesis changes in the first commit might be a bit controversial, and I need to look at the Kafka changes in the second commit a lot more carefully to convince myself I didn't screw something up (1-indexed mzoffset strikes again!)

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6130)

### CODE REVIEW

File: src/dataflow/src/source/kinesis.rs [hunk 1]
```diff
@@ -195,9 +191,8 @@ impl SourceInfo<Vec<u8>> for KinesisSourceInfo {
                             // todo@jldlaughlin: Parse this to determine fatal/retriable?
                             error!("{}", e);
                             self.shard_queue.push_back((shard_id, shard_iterator));
-                            activator.activate();
```
Comment (aljoscha) [inline src/dataflow/src/source/kinesis.rs:223]: Why is it that we can remove these `activate()` calls. The caller of `get_next_message()` seems to also not be calling it.

This is really only for my understanding. 😅

**Comment by aljoscha** [comment]
It seems that `ConsistencyInfo.start_offsets` is very Kafka specific. To me, it's more of a Kafka source configuration than a consistency thing. So maybe as a next step even that could be removed.

File: src/dataflow/src/source/kinesis.rs (hunk 1 above)
Comment (ruchirK) [reply src/dataflow/src/source/kinesis.rs:223]: so now we return `TransientDelay` after which point the caller will call `activate` ( from within the `handle_source` function. its a bit of a 3 step process but i think this is a clearer api for `get_next_message`

**Comment by ruchirK** [comment]
> It seems that ConsistencyInfo.start_offsets is very Kafka specific. To me, it's more of a Kafka source configuration than a consistency thing. So maybe as a next step even that could be removed.

100% agree. I'm doing this work in very small steps because verifying large changes on this part of the codebase is just hard at the moment but I am moving in that direction

**Pull request merged**